src/main/scala/com/fulcrumgenomics/bam/TrimPrimers.scala

```diff
@@ -0,0 +1,195 @@
+  @arg(flag="H", doc="If true, hard clip reads, else soft clip.") val hardClip: Boolean = false,
+  @arg(flag="S", doc="Match to primer locations +/- this many bases.") val slop: Int = 5,
+  @arg(flag="s", doc="Sort order of output BAM file (defaults to input sort order).") val sortOrder: Option[SortOrder] = None
+)extends FgBioTool with LazyLogging {
+  private val mode = if (hardClip) ClippingMode.Hard else ClippingMode.Soft
+
+  Io.assertReadable(input)
+  Io.assertReadable(primers)
+  Io.assertCanWriteFile(output)
+
+  /** A Locatable Amplicon class. */
+  private case class Amplicon(chrom: String, leftStart: Int, leftEnd: Int, rightStart: Int, rightEnd: Int) extends Locatable {
+    def leftPrimerLength: Int    = CoordMath.getLength(leftStart, leftEnd)
+    def rightPrimerLength: Int   = CoordMath.getLength(rightStart, rightEnd)
+    def longestPrimerLength: Int = Math.max(leftPrimerLength, rightPrimerLength)
+
+    override def getContig: String = chrom
+    override def getStart: Int = leftStart
+    override def getEnd: Int = rightEnd
+  }
+
+  override def execute(): Unit = {
+    val in = SamReaderFactory.make().open(input)
+    val outSortOrder = sortOrder.getOrElse(in.getFileHeader.getSortOrder)
+    val outHeader = in.getFileHeader.clone()
+    outHeader.setSortOrder(outSortOrder)
+    val out = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(outHeader, outSortOrder == SortOrder.queryname, output.toFile, null)
+
+    val detector = loadPrimerFile(primers)
+    val maxPrimerLength = detector.getAll.map(_.longestPrimerLength).max
+
+    val iterator = queryNameOrderIterator(in)
+    while (iterator.hasNext) {
+      val reads = nextTemplate(iterator)
+      val rec1 = reads.find(r => r.getReadPairedFlag && r.getFirstOfPairFlag  && !r.isSecondaryOrSupplementary)
+      val rec2 = reads.find(r => r.getReadPairedFlag && r.getSecondOfPairFlag && !r.isSecondaryOrSupplementary)
+
+      (rec1, rec2) match {
+        case (Some(r1), Some(r2)) =>
+          // FR mapped pairs get the full treatment
+          if (!r1.getReadUnmappedFlag && !r2.getReadUnmappedFlag &&
+            r1.getReferenceName == r2.getReferenceName &&
+            SamPairUtil.getPairOrientation(r1) == PairOrientation.FR) {
+
+            val (left, right) = if (r1.getReadNegativeStrandFlag) (r2, r1) else (r1, r2)
+            val (start, end ) = (left.getUnclippedStart, right.getUnclippedEnd)
+            val insert = new Interval(left.getContig, start, end)
+            val amp = detector.getOverlaps(insert).find(amp => abs(amp.leftStart - start) <= slop && abs(amp.rightEnd - end) <= slop)
+            amp match {
+              case Some(amplicon) =>
+                val leftClip  = amplicon.leftPrimerLength
+                val rightClip = amplicon.rightPrimerLength
+                reads.foreach { rec =>
+                  val toClip = if (rec.getFirstOfPairFlag == left.getFirstOfPairFlag) leftClip else rightClip
```

Comment:
It's confusing, but it's doing the right thing.  I'm trying to handle the case where I have either secondary or supplemental records, and I want them all clipped the same as the primary...  So above I figure out which is the `left` and `right` read using position/orientation, and then I go over _all_ reads saying "if you're the same read number as my left read, trim by the `leftClip` amount, else by the `rightClip` amount".  There's an extra level of indirection there.